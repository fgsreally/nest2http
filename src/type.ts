export type RouteInfo={ type: 'Get' | 'Put' | 'Delete' | 'Post', route: string, decorators: string[] }