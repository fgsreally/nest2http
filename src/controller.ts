import { RouteInfo } from "./type"

export class FakeController {
    content: string = ''
    constructor(public name: string, public route: string) { }


    getContent() {
        return this.content
    }

    createMethod(methodName: string, routeInfo: RouteInfo) {

        this.content += `
    ${methodName}(${genParams(routeInfo.decorators)}){
return createRequest(
    {method:'${routeInfo.type.toLowerCase()}',
route:'/${this.route}/${routeInfo.route}',
args:arguments,
params:${JSON.stringify(routeInfo.decorators || [])}}
)
    }
    `

        return this
    }


    createClass() {
        this.content = `
export class ${this.name}{
${this.content}
}`
        return this
    }


    createImport() {
        this.content = `
import {createRequest} from 'nest2http'
`+ this.content
        return this
    }
}


function genParams(decorators: string[]) {

    return decorators.reduce((p, c) => {
        return p +  c+',' 
    }, '')
}