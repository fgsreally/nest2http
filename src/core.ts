import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse } from "axios";

export let instance: AxiosInstance = axios.create({

})

export function createRequest({ params, args, route, method }: {
    method: 'get' | 'post'
    route: string
    params: string[]
    args: any[]
}) {

    if (params.length !== args.length) throw new Error('参数数目不对')
    let queryMap: Record<string, string> = {}

    let body: any

    params.forEach((item, i) => {
        let [type, paraName] = item.split('_')
        switch (type) {
            case 'Param':
                route = resolveParams(route, paraName, args[i]); return;
            case 'Query':
                resolveQuery(queryMap, paraName, args[i]); return
            case 'Body':
                body = args[i]
        }
    })

    const url = resolveUrl(route, queryMap)

    return body ? [method, url, body] : [method, url]

}


function resolveUrl(route: string, query: Record<string, string>) {
    route += '?'
    for (let i in query) {
        if (!route.endsWith('?')) route += '&'
        route += `${i}=${query[i]}`
    }
    return route
}


function resolveQuery(query: Record<string, string>, param: string, arg: string,) {
    query[param] = arg
}


export function resolveParams(route: string, param: string, arg: string) {

    return route.replace(`:${param}`, arg)
}

//@ts-ignore
export declare function useRequest<R>(
    p:R,
    config?: AxiosRequestConfig
): Promise<AxiosResponse<Awaited<R>>>

//@ts-ignore
export function useRequest(reqInfo, config) {
    //@ts-ignore
    return instance[reqInfo.shift()](...reqInfo, config);
}

export function toReq<C extends new (p:any) => any>(controller: C):InstanceType<C> {
        //@ts-ignore
    return new controller()
}


export function injectInstance(ins: AxiosInstance) {
    instance = ins
}