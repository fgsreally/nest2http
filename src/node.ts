export * from './controller'
export * from './plugin'
export * from './transform'


export * from './type'