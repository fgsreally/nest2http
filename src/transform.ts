//@ts-nocheck
import  ts from 'typescript'
import { RouteInfo } from './type'
export function transform(source:string){
    const data: {
        route: string,
        name:string,
        routeMap: Map<string, RouteInfo>
      } = { routeMap: new Map(), route: '' }
      
      ts.transpileModule(source, {
        transformers: { before: [createTransformer()] }
      })
      
      
      
      function createTransformer() {
        return context => {
          return node => ts.visitNode(node, visit)
      
          function visit(node) {

            // 如果发现字符串，替换为自己的内容
            if (ts.isDecorator(node)) {

              if (node.expression?.expression?.escapedText === 'Controller') {
                data.route = node.expression?.arguments?.[0]?.text || ''
                data.name=node.parent.symbol.escapedName
              }
      
      
              if (['Get', 'Put', 'Delete', 'Post'].includes(node.expression?.expression?.escapedText)) {
                data.routeMap.set(node.parent.name?.escapedText, { type: node.expression?.expression?.escapedText, route: node.expression?.arguments?.[0]?.text, decorators: [] })
              }
              if (['Body', 'Query','Param'].includes(node.expression?.expression?.escapedText)) {
                const methodInfo = data.routeMap.get(node.parent?.parent.name?.escapedText)
                methodInfo?.decorators.push(`${node.expression?.expression?.escapedText}_${node.expression?.arguments?.[0]?.text||''}`)
      
      
              }
            }//parent.name?.escapedText
            return ts.visitEachChild(node, visit, context)
      
      
          }
        }
      }
      return data
}