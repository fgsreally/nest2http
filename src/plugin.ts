import { createUnplugin } from 'unplugin'
import { FakeController } from './controller'
import { transform } from './transform'
export const unplugin = createUnplugin(() => {
    return {
        name: 'unplugin-controller-transform',
        enforce:'pre',

        transformInclude(id) {
            return id.endsWith('.controller.ts')
        },
        // just like rollup transform
        transform(code) {
            const ret = transform(code)
            const controller = new FakeController(ret.name, ret.route)
            for (let i of ret.routeMap) {
                controller.createMethod(i[0], i[1])
            }
            return controller.createClass().createImport().getContent()
        },
        // more hooks coming
    }
})
