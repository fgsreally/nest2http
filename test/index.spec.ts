import { FakeController } from "../src/controller";
import { transform } from '../src/transform'
import { createRequest } from "../src/core";
import { describe, expect, it } from 'vitest'
const str = `
import { Controller, Get } from '@nestjs/common';

@Controller('foo')
export class NoVersioningController {
  @Get('/:id')
  helloFoo(@Body()  body:any,@Query('command') cmd:any,@Param('id') param:string) {
    return 'Hello FooBar!';
  }
  
}`

describe('nest2http', () => {
    it('transform nest controller to fake controller', () => {
        const ret = transform(str)
        expect(ret.name).toBe('NoVersioningController')
        expect(ret.route).toBe('foo')
        const controller = new FakeController(ret.name, ret.route)
        for (let i of ret.routeMap) {
            controller.createMethod(i[0], i[1])
        }
        const content = controller.createClass().createImport().getContent()
        expect(content).toMatchSnapshot()
    })

    it('createRequest', () => {
        class fakeClass {

            helloFoo(Body_: any, Query_command: any, Param_id: any,) {
                return createRequest(
                    {
                        method: 'get',
                        route: '/foo/:id',
                        args: arguments,
                        params: ["Body_", "Query_command", "Param_id"]
                    } as any
                )
            }

        }
        const { helloFoo } = new fakeClass()
        expect(helloFoo({ name: 'nest2http' }, 'cmd', 'param')).toMatchSnapshot()
    })
})