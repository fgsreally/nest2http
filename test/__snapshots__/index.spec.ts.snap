// Vitest Snapshot v1

exports[`nest2http > createRequest 1`] = `
[
  "get",
  "/foo/param?command=cmd",
  {
    "name": "nest2http",
  },
]
`;

exports[`nest2http > transform nest controller to fake controller 1`] = `
"
import {createRequest} from 'nest2http'

export class NoVersioningController{

    helloFoo(Body_,Query_command,Param_id,){
return createRequest(
    {method:'get',
route:'/foo/:id',
args:arguments,
params:[\\"Body_\\",\\"Query_command\\",\\"Param_id\\"]}
)
    }
    
}"
`;
